### ISSUE TITLE
TimeSeries

### ISSUE THREAD

**Issue by BorisHollas** [wookietreiber/scala-chart#23]
Does scala-chart provide a wrapper for JFreeChart TimeSeries?

**Comment by wookietreiber** [comment]
Not a wrapper per se, but you can convert from any Scala collection to a JFreeChart dataset via `toTimeSeries`, `toTimeSeriesCollection`, `toTimePeriodValues` and `toTimePeriodValuesCollection`.

You will need some sort of time values for the domain, though, i.e. values that can convert implicitly to a JFreeChart `RegularTimePeriod` for the `TimeSeries*` conversions or `TimePeriod` for the `TimeValues*` conversions. This means that either the collection you *convert from* already has e.g. a [JFreeChart `Second`](http://www.jfree.org/jfreechart/api/javadoc/org/jfree/data/time/Second.html) as its domain type, then you won't need an implicit conversion, or you have something else like [`joda.time.DateTime`](http://www.joda.org/joda-time/apidocs/org/joda/time/DateTime.html) then you will need an implicit like the following (imports, etc. excluded for brevity):

```scala
implicit def joda2jfreeminute(d: DateTime): Minute = ???
```

All chart factories provided by scala-chart, e.g. `XYLineChart`, will use a `DateAxis` if you supply one of the abovementioned JFreeChart datasets.

**Comment by BorisHollas** [comment]
So, if I have an Array[(java.util.Date, Double)] or a List[Array[(java.util.Date, Double)] ], can I use toTimeSeries or toTimeSeriesCollection to convert this to the corresponding JfreeChart type?

Also, if I have a TimeSeriesCollection where all time ranges are equal (e.g., from 2015-1-1 until now), how can I produce an XYLineChart that stacks the data with a common x-axis and how can I specify the vertical size of each subplot?

**Comment by wookietreiber** [comment]
```
import language.implicitConversions
import java.util.Date
import org.jfree.data.time.Second
import org.jfree.data.time.RegularTimePeriod
import scalax.chart._
import scalax.chart.api._

implicit def jdate2jfree(d: Date): RegularTimePeriod = new Second(d)

val data = List[(Date,Double)]()
val data2 = List[(String,List[(Date,Double)])]()

data.toTimeSeries("series name")
data2.toTimeSeriesCollection
```

For the `TimeSeriesCollection` you need an identifier for each series, in this case `String`, i.e. how each series should be named (you'll see the series names in the chart legend). If you put such a dataset into a chart, the different series' get represented by colored lines/areas/bars.

What exactly do you mean by *vertical size of each subplot*?

**Comment by BorisHollas** [comment]
I want each time series to appear in a frame of its own, with a common x-axis, as in this example:[https://1.bp.blogspot.com/-ZuCBGy0rYwU/T_xIBulGULI/AAAAAAAAEko/QRKpPpGyQXs/s1600/Differential+(3).png](url)

**Comment by wookietreiber** [comment]
I don't have anything in scala-chart for this specific use case. My guess is, that you need to take a look at [CombinedRangeXYPlot](http://www.jfree.org/jfreechart/api/javadoc/org/jfree/chart/plot/CombinedRangeXYPlot.html).

**Comment by wookietreiber** [comment]
I think, though, I can add such combined range/domain factories. I'll get back to you on that.

**Comment by wookietreiber** [comment]
A factory for combined domain plots:

```scala
      def domain[A: ToXYDataset](data: A*)(
        title: String = "",
        orientation: Orientation = Orientation.Vertical,
        legend: Boolean = true)
        (implicit theme: ChartTheme = ChartTheme.Default): XYChart = {

        val datasets = data.map(ToXYDataset[A].convert)

        val subplots = datasets map { dataset =>
          val dateAxis = needsDateAxis(dataset)
          val domainAxis = XYDomainAxis(dateAxis)
          val rangeAxis = new NumberAxis()
          val renderer = new XYLineAndShapeRenderer(true, false)
          val plot = new XYPlot(dataset, domainAxis, rangeAxis, renderer)
          plot.setOrientation(orientation)
          plot
        }

        val plot = new CombinedDomainXYPlot()

        subplots foreach plot.add

        XYChart(plot, title, legend, theme)
      }
```

Fill with data:

```
def data = for (i <- 0 to 10) yield i -> util.Random.nextDouble
def series(name: String) = data.toXYSeries(name)
val chart = XYLineChart.combined.domain(List.tabulate(3)(x => series(x.toString)))()
chart.show()
```

This looks like that:

![combined-not-split](https://cloud.githubusercontent.com/assets/827088/16192488/ef5eb7d8-36e9-11e6-932d-d470f8ca0355.png)

What I don't know yet is how to split the axis into multiple parts like in the picture you provided. As far as I can see there is no such axis class in JFreeChart.

**Comment by BorisHollas** [comment]
On page 98 of the manual is an example with CombinedDomainXYPlot and two subplots as in my picture above. However, the domain axis is null in the example.

**Comment by wookietreiber** [comment]
I made a mistake, my own dataset conversions trolled me and were converting the `List[XYSeries]` to a single `XYSeriesCollection` instead of to a `List[XYSeriesCollection]`.

The fill with data section must actually look like:

```scala
def data = for (i <- 0 to 10) yield i -> util.Random.nextDouble
def series(name: String) = data.toXYSeriesCollection(name)
def datasets: List[XYSeriesCollection] = List.tabulate(3)(x => series(x.toString))
XYLineChart.combined.domain(datasets: _*)().show()
```

Then it actually works:

![foo](https://cloud.githubusercontent.com/assets/827088/16225276/219e517e-37a6-11e6-845b-91b41982df0d.png)

I'll add factories for the combined range/domain charts later, but you can already use the chart factory I have included in my previous comment.

**Comment by BorisHollas** [comment]
That's fine, thanks! The plot.add method also has an optional second parameter that defaults to 1 and specifies the relative vertical size of the subplot.

**Comment by wookietreiber** [comment]
Yes, I have also experimented a little with the subplot weight. For a generic factory I would leave it at `1` for all subplots. Users can then later configure it in detail via `plot.setWeight`.

**Comment by BorisHollas** [comment]
Does domain work with data: TimeSeriesCollection and what do I have to import to make the code snippet compile.

**Comment by wookietreiber** [comment]
Sorry for not replying sooner. I am currently in the process of reorganizing the chart factories. The reason for this reorganization is that with the addition of the combined domain **and** range factories, the number of factories is increasing significantly. I am looking into ways to minimize this increase. Once I am finished with this reorganization I will release a new scala-chart version.

---

Until the release, here is a complete example you can use. Just replace the `data`, `dataset` and/or `datasets` function with how you generate your data.

```scala
package org.example

import language.implicitConversions

import java.util.Date

import org.jfree.chart.axis.{DateAxis, NumberAxis}
import org.jfree.chart.plot.CombinedDomainXYPlot
import org.jfree.chart.renderer.xy.XYLineAndShapeRenderer
import org.jfree.data.time.{RegularTimePeriod, Second}

import scalax.chart._
import scalax.chart.api._

object CombinedDomain extends App {
  implicit def jdate2jfree(d: Date): RegularTimePeriod = new Second(d)

  def data: Seq[(Date,Double)] = for (i <- 1000 to 10000 by 1000) yield
    new Date(i.toLong) -> util.Random.nextDouble

  def dataset(name: String) = data.toTimeSeriesCollection(name)

  def datasets = List.tabulate(3)(x => dataset(x.toString))

  def combinedDomain(datasets: TimeSeriesCollection*)
           (implicit theme: ChartTheme = ChartTheme.Default): XYChart = {

    val subplots = datasets map { dataset =>
      val rangeAxis = new NumberAxis()
      val renderer = new XYLineAndShapeRenderer(true, false)
      val plot = new XYPlot(dataset, null, rangeAxis, renderer)
      plot
    }

    val domainAxis = new DateAxis()

    val plot = new CombinedDomainXYPlot(domainAxis)

    subplots foreach plot.add

    XYChart(plot, title = "", legend = true, theme)
  }

  val chart = combinedDomain(datasets: _*)

  chart.show()
}
```
